chore: create ts/nextjs/express shell

portenez opened the pull request:
This is intended to be merged after #2 

**Describe your changes**
Init minimal project shell and linters (no ci actions for now)

**Testing performed**

- `npm run dev` works locally
- `npm eslint` works locally
- `npm stylelint` works locally
- `npm test` works locally



**Additional context**
There will be another PR for CI automation for the linting and unit tests

portenez committed:
chore: create ts/nextjs/express/gql shell

Signed-off-by: vgarciavalen <info@vgarcia.dev>

dakotalillie commented on .eslintignore:
```diff
@@ -0,0 +1,3 @@
+dist/
+!.storybook
```
Are we using [storybook](https://storybook.js.org)?

portenez replied:
Not yet, but I was planning to add it later. 

Removing this makes sense until we actually add storybook. However, leaving it is not harmful either. 

What's your pref?

dakotalillie replied:
I'm open to leaving it if we're eventually going to use it. My understanding is it's usually intended for component libraries — would we use it here to preview components in isolation before adding them to the frontend proper?

dakotalillie commented on package.json:
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "coffechat",
```
At this point, I'm unsure if `coffechat` is just a typo that has popped up repeatedly, or it's actually the name of the project 😜

portenez replied:
Oh, man. It's a quirk of my Spanglish brain. Cafe/coffee merge in my mind, so I keep repeating the same mistake. I def, should fix.

dakotalillie commented on .eslintrc.yml:
```diff
@@ -0,0 +1,92 @@
+  # NOTE: any rule set to warning (warn or 1)
+  # eslint will NOT  output any warning when running `npm test` or `npm run eslint`
+  # warnings will still appear on the editor 
+  no-nested-ternary: 0
```
I would personally prefer not to use nested ternaries. Do you find them useful?

portenez replied:
Yeah, I do find them useful, but I see the harm. I'm totally ok with disallowing.

dakotalillie commented on .eslintrc.yml:
```diff
@@ -0,0 +1,92 @@
+  quotes:  
+    - error
+    - single # has to match prettier quotes
+    - avoidEscape: true # has to always be true to not clash with preetier
```
I'm curious why we need formatting rules like this for eslint. I'd imagine they'd fall under the purview of prettier.

```suggestion
    - avoidEscape: true # has to always be true to not clash with prettier
```

dakotalillie commented on .gitignore:
```diff
@@ -0,0 +1,122 @@
+.nuxt
+dist
+
+# Gatsby files
```
**Nitpick**

Most of what's in this `.gitignore` does not seem to be used in this project.

portenez replied:
I used https://www.toptal.com/developers/gitignore to generate the `.gitignore` file.

Lately, I've been opting to do this because I've been burned by crafting the `.gitignore` manually. Sometimes, I forget (or don't even know I have to) add files to `.gitignore` and end up committing something that I shouldn't.

Lemme know what you think after my explanation. I'm fine with cleaning this up.

dakotalillie replied:
Eh. It's a major nitpick. Ain't doin no harm.

dakotalillie commented on .prettierrc.yml:
```diff
@@ -0,0 +1,4 @@
+trailingComma: "es5"
+semi: false 
+singleQuote: true
+tabWith: 2
```
Willing to add `proseWrap: always`, by any chance? And maybe bump `printWidth` up to 100 or 120?

portenez replied:
for `printWidth` my only problem is that I constantly split my editors vertically in `vscode`.  I also often put a terminal and vscode in the same monitor splitted vertically across the window 

- 120 is definitely too long and will give me a lot of hassle.
- 100 is a bit better, but still gives me some hassle. 

See screenshot:
![image](https://user-images.githubusercontent.com/1408006/116808144-3c501700-ab05-11eb-9d06-3fef9ebdfd5b.png)

dakotalillie replied:
Fair enough — keeping `printWidth` at 80 is fine. Any chance for `proseWrap`?

dakotalillie replied:
Ah sorry, I see now that you added it. Was using the wrong view.

dakotalillie commented on README.md:
```diff
@@ -18,6 +18,10 @@ scheduling appointments as easy as possible.
 ## Quick Start
 
+1. Clone this repository
+1. Run `npm init`
```
Why do we need to run `npm init`? This project already has a `package.json`

dakotalillie commented on README.md:
```diff
@@ -18,6 +18,10 @@ scheduling appointments as easy as possible.
 
+1. Clone this repository
+1. Run `npm init`
+1. Start the service by `npm run dev`
```
Perhaps worth adding a note explaining more thoroughly what this does?

portenez replied:
What do you have in mind? Maybe, add the note the way you think would be helpful?

dakotalillie commented on doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md:
```diff
@@ -0,0 +1,37 @@
+to aid in our development.
+
+- [express.js](https://expressjs.com) is the most used `nodejs` web framework.
+- [React.js](https://reactjs.org/) is well-known and widely use `js` library
```
```suggestion
- [React.js](https://reactjs.org/) is well-known and widely used `js` library
```

dakotalillie commented on doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md:
```diff
@@ -0,0 +1,37 @@
+- [express.js](https://expressjs.com) is the most used `nodejs` web framework.
+- [React.js](https://reactjs.org/) is well-known and widely use `js` library
+  for building component-based user interfaces.
+- [styled-components](https://styled-components.com/) is modern `css-in-js`
```
I might prefer [Emotion](https://emotion.sh/docs/introduction) to Styled Components — they have very similar APIs at this point, but I've found Emotion to be a bit more reliable and easier to set up. What do you think?

dakotalillie commented on doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md:
```diff
@@ -0,0 +1,37 @@
+  framework
+- [Next.js](https://nextjs.org/) is a framework to "outsource" the complexity
+  of running a React app.
+- [eslint](https://eslint.org/) is a static analysis tools to find bugs
```
```suggestion
- [eslint](https://eslint.org/) is a static analysis tool to find bugs
```

dakotalillie commented on doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md:
```diff
@@ -0,0 +1,37 @@
+- [eslint](https://eslint.org/) is a static analysis tools to find bugs
+  and anti-patterns in javascript (and typescript) code.
+- [stylelint](https://stylelint.io/) helps you avoid errors and enforce
+  conventions in your styles works with styled-components out of the box.
```
```suggestion
  conventions in your styles, and works with styled-components out of the box.
```

dakotalillie commented on doc/architecture/decisions/0003-use-nextjs-express-stylecomponents.md:
```diff
@@ -0,0 +1,37 @@
+- [stylelint](https://stylelint.io/) helps you avoid errors and enforce
+  conventions in your styles works with styled-components out of the box.
+- [prettier](https://prettier.io/) is an opinionated code formatter to avoid
+  spending values development time agreeing on a code style.
```
```suggestion
  spending valuable development time agreeing on a code style.
```

dakotalillie commented on package.json:
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "coffechat",
+  "version": "0.0.0",
+  "description": "Connect Mentors and Mentees by making the process of connecting and then scheduling appointments as easy as possible.",
```
Could probably use a catchier description here. @pooja-malpani do you have anything on hand?

dakotalillie commented on package.json:
```diff
@@ -0,0 +1,71 @@
+    "eslint": "eslint .",
+    "eslint:fix": "eslint . --quiet --fix",
+    "stylelint": "stylelint '**/*.tsx'",
+    "stylelint:fix": "stylelint '**/*.tsx' --fix"
```
I'd imagine we probably need `prettier` and `prettier:fix` commands here as well, no?

portenez replied:
So, the way this is set up, `esllint` will run prettier as part `eslint` and `eslint:fix`. So we're good on the javascript files.

For the markdown, yaml, etc. I was not planning/requiring on running prettier, but that's useful.  do you think I should do this?

dakotalillie replied:
Gotcha. I personally like using Prettier on markdown in particular

dakotalillie commented on src/pages/_document.tsx:
```diff
@@ -0,0 +1,33 @@
+import React from 'react'
```
Correct me if I'm wrong, but since we're using React 17, I don't believe it's necessary to `import React from 'react'` anymore, here or elsewhere.

portenez replied:
Needed to update an eslint rule. Should be good now.

dakotalillie commented on src/lib/Config.ts:
```diff
@@ -0,0 +1,116 @@
+  )
+}
+
+const Config = <T extends Spec>(
```
Why is this capitalized, if it's not a class, component, or factory function?

dakotalillie commented on src/lib/Config.ts:
```diff
@@ -0,0 +1,116 @@
+type SpecValue =
```
Personally, I find this file rather confusing. It seems like it is parsing, validating, and providing overrides and default values all at once. Am I understanding it correctly? If so, what would you say to splitting those things up into separate, single-purpose functions which can be chained?

dakotalillie replied:
I rewrote this using [ajv](https://ajv.js.org)

dakotalillie commented on src/pages/_document.tsx:
```diff
@@ -0,0 +1,33 @@
+import Document, { DocumentContext, DocumentInitialProps } from 'next/document'
+import { ServerStyleSheet } from 'styled-components'
+
+export default class MyDocument extends Document {
```
Could we use a better name for this than `MyDocument`, or add a comment to indicate why a custom `Document` is necessary?

portenez replied:
ah, yeah.  I copied it over from the docs. Will add a comment with a link.

As for the name, can you help me with a suggestion? I'm blanking on a good name: `CustomDoc` is as bad as `MyDocument` IMO. Then, there's only one single `_document.tsx` in next js, so we cannot make it super specific like `StyledComponentsBootstrapDoc` or something like that, because if we ever change it won't make sense.

dakotalillie replied:
changed to `DocumentWithStyles`. Let me know if you'd prefer something different.

dakotalillie commented on tsconfig.json:
```diff
@@ -0,0 +1,45 @@
+  },
+  "exclude": [
+    "node_modules",
+    "typings",
```
What is `typings`?

portenez replied:
ah, it's an old school way of defining the `typescript types`. There use to be competitor to `definitely` types used `typings` instead of `@types`. I'll remove

dakotalillie commented on tsconfig.server.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "lib": ["ES2019"],
+    "target": "ES2019",
```
Would esnext work here?

```suggestion
    "lib": ["esnext"],
    "target": "esnext",
```

portenez committed:
chore: address various PR issues

Co-authored-by: Dakota Lillie <dakota.lillie@icloud.com>

- build(eslint): disallow nested ternaries
- build(eslint): fix typo
- build(prettier): set proseWrap to true
- docs(README): fix wrong instructions
- chore: enable no react import

Signed-off-by: vgarciavalen <info@vgarcia.dev>

portenez reviewed (commented):
I made some corrections and addressed some comments. I will try to address these ones soon. 

In other words, I wasn't able to go through all the comments in this pass.

Also, if anyone has any time to help me finalize this, please do help me. I mean add to this PR branch.

portenez committed:
chore(nextjs): add custom doc comment

Signed-off-by: vgarciavalen <info@vgarcia.dev>

dakotalillie commented:
I'll see about pushing up some commits to address my own suggestions.

portenez commented:
So I pulled the changes from master using the github ui, and that might create a messy commit history. Before merging let's see if we can cleanup the history (squash merge is disabled in this repo)

dakotalillie commented:
> squash merge is disabled in this repo

😱

portenez commented:
I'm in the car (wife driving). I resolved the storybook comment, and now can't get it back easily (in mobile ui).

But storybook is good for not only to preview but also for:

- writing simple snapshot tests at the same time the stories are written
- paralellize work. Devs can start working on components while backend work is still not ready.
- iterate with ux quickly (the storybook is easy to deploy to personal github pages)

dakotalillie commented:
Righto. Sounds good to me 👍

dakotalillie committed:
fix: typos and formatting in adr

Signed-off-by: Dakota Lillie <dakota.lillie@icloud.com>

dakotalillie committed:
refactor: remove unnecessary import of react

Signed-off-by: Dakota Lillie <dakota.lillie@icloud.com>

dakotalillie committed:
refactor: rename to DocumentWithStyles

Signed-off-by: Dakota Lillie <dakota.lillie@icloud.com>

dakotalillie committed:
chore: tsconfig updates

Signed-off-by: Dakota Lillie <dakota.lillie@icloud.com>

portenez commented:
Thank you. Everything looks great. The only thing to double check is the target for tsconfig server. That one needs to match the nodejs version that we're targeting.

portenez commented:
Whoever gets a chance, let's update the tsconfig version according to this
https://github.com/tsconfig/bases#centralized-recommendations-for-tsconfig-bases

dakotalillie committed:
refactor: use ajv for object parsing

Signed-off-by: Dakota Lillie <dakota.lillie@icloud.com>

portenez commented on src/server/env.ts:
```diff
@@ -0,0 +1,24 @@
+  PORT: number
+}
+
+const parse = makeParser<Env>({
```
very cool! So, pretty much I had reimplemented this functionality.

portenez commented on src/server/env.ts:
```diff
@@ -0,0 +1,24 @@
+  required: [],
+})
+
+export default parse((process.env as unknown) as Env)
```
**nitpick**:

ok, this might be wishful thinking... but is there any way we could somehow export this in a way we could import like:

```typescript
import { PORT } from './env'
```

this is already pretty cool, but this would be very convenient.

portenez replied:
ok, I'm just wondering.. `process.env` interface is actually `Record<string, string>`. Would this affect at all this implementation? Could this make the use of this a bit confusing?

Also, we might be able to change the types in the `makeParser` funct, to make it clearer.

portenez commented on src/server/index.ts:
```diff
@@ -1,13 +1,9 @@
 import log, { final } from './log'
 import createServer from './createServer'
-import Config from '../lib/Config'
+import env from './env'
```
Per my other comment: https://github.com/bloomberg/coffeechat/pull/3/commits/92b506e67f2b647376cfb85836ea3e17ae7bc480#r624785942



I'm wondering if we could somehow export the properties. This is already pretty, good. I'm just wondering.

portenez committed:
refactor(makeparser): cleanup types

Signed-off-by: vgarciavalen <info@vgarcia.dev>

portenez commented on src/lib/makeParser.spec.ts:
```diff
@@ -0,0 +1,109 @@
+import makeParser, { ParseError } from './makeParser'
```
I have no hard pref, but I tend to use `__test__/*.spec.ts` for the tests. What standard should we use?
with or without `__test__` ?

dakotalillie replied:
I tend to like to put the spec files directly next to what they're testing, because in general I like to 1. colocate code as much as possible, and 2. use flat structures as much as possible. However, I do think it's just a matter of personal preference — as the project lead, if you'd prefer to use `__test__`, I'll defer to you.

portenez committed:
refactor(env): export env props

Signed-off-by: vgarciavalen <info@vgarcia.dev>

portenez committed:
build(engines): enforce node <= 14 < 15 and npm 7

Signed-off-by: vgarciavalen <info@vgarcia.dev>

portenez commented on src/server/env.ts:
```diff
@@ -0,0 +1,26 @@
+  required: [],
+})
+
+const { LOG_LEVEL, LOG_PRETTY, NODE_ENV, PORT } = parse(process.env)
```
@dakotalillie What do you think of this? it's a bit verbose... but gives me the convenience I was looking for lol. You're either verbose here, or later.

dakotalillie replied:
I think it's a good idea to destructure all the properties in here! Updated to make it a bit more concise, but I like the move in general.

portenez committed:
build(prettier): enable more prettier

Signed-off-by: vgarciavalen <info@vgarcia.dev>

portenez commented on README.md:
```diff
@@ -29,44 +33,40 @@ scheduling appointments as easy as possible.
 
 We :heart: contributions.
 
-Have you had a good experience with this project? Why not share some
```
I ran prettier on everything. That's why the changes. please ignore whitespace for a better experience.

portenez commented:
@dakotalillie  I pushed a few things:

- Better types for `makeParser`
- My wish of being able to use `import { PORT } from './env'
- `.vscode` with standardized formatting, recommended plugins and such (there are pros and cons to this, lemme know what do you think). At least the recommendations should be there. As for the settings, It's useful for people to immediately get started with a working config. 
- I ran prettier on everything and added prettier targets.

Let's try to merge this soon. 
As for the `emotion`, let's do that as a quick follow-up? I have used emotion before but never with nextjs. My pet peeve with emotion is having to do this for the global styles:
https://emotion.sh/docs/globals

Lol, obviously it's a nitpick. So, in sum yeah I'm good with emotion.

dakotalillie committed:
refactor: concise env exports

Signed-off-by: Dakota Lillie <dakota.lillie@icloud.com>

dakotalillie committed:
refactor: remove unnecessary type assertions

Signed-off-by: Dakota Lillie <dakota.lillie@icloud.com>

dakotalillie committed:
chore: .vscode tweaks

Signed-off-by: Dakota Lillie <dakota.lillie@icloud.com>

dakotalillie commented:
@portenez those type improvements do make it a lot cleaner — thank you for that.

I'm un-opinionated regarding adding the `.vscode` settings. I tend not to add them myself, but I am open-minded about it being a beneficial move which I just haven't explored yet. Just made a couple minor tweaks, including changing the extension for `settings.json` to `.jsonc` since you have a comment in there.

I've just been in a couple situations where Styled Components has inadvertently released a breaking change in a minor version bump, or where I struggled to get it working so I switched to Emotion which worked immediately. There were also some comparisons a while back about how Emotion was smaller and performed better than Styled Components, although I think that gap may have been bridged.

I think these are all nitpicks — I'm good with Styled Components too. Just wanted to bring up the point and have a small conversation. Ultimately, as with the conversation around using `__test__` directories, I'll defer to you.

dakotalillie committed:
docs: readme and description updates

Signed-off-by: Dakota Lillie <dakota.lillie@icloud.com>

dakotalillie committed:
chore: change server target to es2020

Signed-off-by: Dakota Lillie <dakota.lillie@icloud.com>

dakotalillie commented:
Changed the server target to ES2020 given the [Node Target Mappings](https://github.com/microsoft/TypeScript/wiki/Node-Target-Mapping) and the fact that the `.nvmrc` and `package.json` both say to use Node 14.

dakotalillie reviewed (dismissed):
Feel free to make any additional changes based on our discussions around placement of tests and potential usage of Emotion. This all looks good to me 🎉

portenez committed:
test(jest): allow only 1 file pattern

Signed-off-by: vgarciavalen <info@vgarcia.dev>

portenez committed:
docs(readme): instruct to use sample.env

Signed-off-by: vgarciavalen <info@vgarcia.dev>

portenez commented:
@dakotalillie Can you review/approve again? I did a few things and the approval was dismissed.

Here is what I did:

- removed the `merge` commit from `main` to get a (slightly) cleaner git history
- rebased from `main`
- added two commits:
  - Small updates to the README
  - Switched to using `__tests__` , and only configuring `jest` to only allow that pattern.

portenez committed:
build(css): switch to emotion

Signed-off-by: vgarciavalen <info@vgarcia.dev>

portenez commented:
Also, I switched this to emotion! 🧑‍🎤

Commit:
chore: minor tweaks

Signed-off-by: Dakota Lillie <dlillie2@bloomberg.net>

dakotalillie reviewed (approved):
LGTM 🎉

dakotalillie commented:
Last commit isn't verified cuz I'm using my work laptop at the moment and haven't set up a GPG key on it yet. Will do that shortly — let me know if you'd like me to re-push the commit.

portenez commented:
Given we cannot use `squash merge`, let's use the default github merge comment. 
Tools like semantic-release, usually accept those (even if they're not conventional-commits)
